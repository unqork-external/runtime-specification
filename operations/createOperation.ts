import type { OperationTypes } from './enums/operation-types.enum'
import { OperationTypeMap } from './typemaps/operation.typemap'
import type { CustomOperationTypeMap } from '../../operations'
import type { InputRef } from '../inputs/inputRef'

/**
 * Given an OperationType, Typescript will complete the options list.
 *
 * In return, the caller receives Typescript constructed class.
 *
 * This includes preset defaults.
 *
 * In essence, this function ensures that all keys are fully present, fully typed, and ready for runtime
 * validation if necessary.
 *
 * @param type
 * @param options
 * @param name name of the operation
 * @param additional
 * @param spreadOp
 * @param additionalArgs
 * @returns - The Operation Class type
 */
export const createOperation = <OpType extends OperationTypes>(
  type: OpType,
  options: InstanceType<(typeof OperationTypeMap)[OpType]>['options'],
  additionalArgs?: Partial<AdditionalArgs>,
): InstanceType<(typeof OperationTypeMap)[OpType]> => {
  const classOpType = OperationTypeMap[type]
  const op = new classOpType()
  return Object.assign(op, {
    type,
    options: {
      ...op.options,
      ...options,
    },
    ...additionalArgs,
  })
}

export const createCustomOperation = <OperationEnum extends string, OpType extends OperationEnum>(
  type: OpType,
  options: InstanceType<CustomOperationTypeMap<OperationEnum>[OpType]['model']>['options'],
  typeMap: CustomOperationTypeMap<OperationEnum>,
  additionalArgs?: Partial<AdditionalArgs>,
) => {
  const classOpType = typeMap[type].model
  const op = new classOpType()
  return Object.assign(op, {
    type,
    options: {
      ...op.options,
      ...options,
    },
    ...additionalArgs,
  })
}

interface AdditionalArgs {
  name: string
  notifyImmediately: boolean
  creatorSummary: string
}
