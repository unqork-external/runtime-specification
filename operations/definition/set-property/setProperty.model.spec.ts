import { SetPropertyOperation } from './setProperty.model'
import { generateSchemaAndValidate } from '../../../../utilities'
import { createOperation } from '../../createOperation'
import { OperationTypes } from '../../enums/operation-types.enum'

describe('Set Property Model', () => {
  let validate, schema
  beforeAll(() => {
    const response = generateSchemaAndValidate(SetPropertyOperation)
    validate = response.validate
    schema = response.schema
  })

  it('should match this JSON Schema', () => {
    expect(schema).toStrictEqual({
      type: 'object',
      properties: {
        type: {
          type: 'string',
          const: 'SET_PROPERTY',
          examples: ['SET_PROPERTY'],
        },
        options: {
          $ref: '#/definitions/SetPropertyOptions',
        },
        name: {
          description: 'Name of the operation',
          type: 'string',
        },
      },
      required: ['options'],
      definitions: {
        SetPropertyOptions: {
          type: 'object',
          properties: {
            targetKey: {
              type: 'string',
              description:
                '`targetKey` refers to the intended target to manipulate. Can work on an array of targetKeys as well.',
              examples: ['firstNameTextField', 'grid.row(0).col(0)'],
              minLength: 1,
            },
            property: {
              type: 'string',
              description:
                '`property` refers to the intended property to mutate. This can be represented as a JSONPath expression.',
              examples: ['value', 'validation.latest'],
              minLength: 1,
            },
            value: {
              type: ['null', 'integer', 'number', 'string', 'boolean', 'array', 'object'],
              description: 'The value refers to the intended value to mutate the intended property to.',
            },
            shouldOverwrite: {
              type: 'boolean',
              description:
                //eslint-disable-next-line max-len
                "shouldOverwrite refers to the type of behavior that SET_PROPERTY should emulate. If true, SET_PROPERTY will overwrite the value located at the intended property of the target regardless of if the intended property is set. If false, SET_PROPERTY will check if the property's value is set, and if so, SET_PROPERTY will exit.",
              default: true,
            },
          },
          description:
            //eslint-disable-next-line max-len
            'SET_PROPERTY will mutate a specific property on a specific target to a specific value. Optionally, if shouldOverwrite is set to false and if a value is already existing on the targeted property, then the Operation will do nothing and exit.',
          required: ['targetKey', 'property'],
        },
      },
    })
  })

  it('Set Property should validate appropriately', () => {
    const op = createOperation(OperationTypes.SET_PROPERTY, {
      targetKey: 'name1',
      value: 'test',
      property: 'value',
    })

    const isValid = validate(op)
    expect(isValid).toBeTrue()
  })

  it('Ensure shouldOverwrite is set to default - also ensure optional keys are set.', () => {
    const op = createOperation(OperationTypes.SET_PROPERTY, {
      targetKey: 'name1',
      value: 'test',
      property: 'value',
    })

    expect(op.options.shouldOverwrite).toBe(true)
  })

  it('should error if targetKey is blank', () => {
    const op = createOperation(OperationTypes.SET_PROPERTY, {
      targetKey: '',
      value: 'something',
      property: 'value',
    })

    const isValid = validate(op)
    expect(isValid).toBeFalse()
  })

  it('should error if property is blank', () => {
    const op = createOperation(OperationTypes.SET_PROPERTY, {
      targetKey: 'firstName',
      value: 'something',
      property: '',
    })

    const isValid = validate(op)
    expect(isValid).toBeFalse()
  })

  it('should not error if value is blank', () => {
    const op = createOperation(OperationTypes.SET_PROPERTY, {
      targetKey: 'firstName',
      value: '',
      property: 'property',
    })

    const isValid = validate(op)
    expect(isValid).toBeTrue()
  })

  it('should not error if value is null', () => {
    const op = createOperation(OperationTypes.SET_PROPERTY, {
      targetKey: 'firstName',
      value: null,
      property: 'property',
    })

    const isValid = validate(op)
    expect(isValid).toBeTrue()
  })

  it('should not error if value is undefined', () => {
    const op = createOperation(OperationTypes.SET_PROPERTY, {
      targetKey: 'firstName',
      value: undefined,
      property: 'property',
    })

    const isValid = validate(op)
    expect(isValid).toBeTrue()
  })
})
