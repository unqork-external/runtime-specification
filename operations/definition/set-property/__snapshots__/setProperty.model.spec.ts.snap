// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Set Property Model should match this JSON Schema 1`] = `
{
  "definitions": {
    "InputRef": {
      "description": "Inputs can be used to utlize dynamic values from other state within the runtime
  within a particular execution of an operation",
      "properties": {
        "alias": {
          "description": "Optionally alias resolved value for use in formulas",
          "type": "string",
        },
        "propertyPath": {
          "description": "Path to property on target to resolve",
          "type": "string",
        },
        "required": {
          "description": "If true, operation won't execute if targeted property has no value",
          "type": "boolean",
        },
        "silent": {
          "description": "Not used in later formulas, only used to determine if operation should execute",
          "type": "boolean",
        },
        "targetKey": {
          "description": "Key of target this input depends on",
          "minLength": 1,
          "type": "string",
        },
      },
      "required": [
        "targetKey",
      ],
      "type": "object",
    },
    "SetPropertyOptions": {
      "description": "SET_PROPERTY will mutate a specific property on a specific target to a specific value. Optionally, if shouldOverwrite is set to false and if a value is already existing on the targeted property, then the Operation will do nothing and exit.",
      "properties": {
        "inputs": {
          "description": "
    Inputs to be used within formulas in other operation options
  ",
          "items": {
            "$ref": "#/definitions/InputRef",
          },
          "type": "array",
        },
        "property": {
          "description": "\`property\` refers to the intended property to mutate. This can be represented as a JSONPath expression.",
          "examples": [
            "value",
            "validation.latest",
          ],
          "minLength": 1,
          "type": "string",
        },
        "resolveSyntax": {
          "description": "Temporary way to resolve a value as syntax. Currently only used for auto generation of CSMs. Runtime team to revisit.",
          "type": "boolean",
        },
        "shouldExecute": {
          "description": "
    Conditionally execute this operation based on a formula.
    Can set to false to temporarily disabled when debugging
  ",
          "type": [
            "string",
            "boolean",
          ],
        },
        "shouldOverwrite": {
          "default": true,
          "description": "shouldOverwrite refers to the type of behavior that SET_PROPERTY should emulate. If true, SET_PROPERTY will overwrite the value located at the intended property of the target regardless of if the intended property is set. If false, SET_PROPERTY will check if the property's value is set, and if so, SET_PROPERTY will exit.",
          "type": "boolean",
        },
        "targetKey": {
          "description": "\`targetKey\` refers to the intended target to manipulate. Can work on an array of targetKeys as well.",
          "examples": [
            "firstNameTextField",
            "grid.row(0).col(0)",
          ],
          "minLength": 1,
          "type": "string",
        },
        "value": {
          "description": "The value refers to the intended value to mutate the intended property to.",
          "type": [
            "null",
            "integer",
            "number",
            "string",
            "boolean",
            "array",
            "object",
          ],
        },
      },
      "required": [
        "targetKey",
        "property",
      ],
      "type": "object",
    },
  },
  "description": "The Set Property operation sets a value at a property location for a given target.",
  "properties": {
    "creatorSummary": {
      "description": "A detailed summary of the operation",
      "type": "string",
    },
    "name": {
      "description": "Name of the operation",
      "type": "string",
    },
    "options": {
      "$ref": "#/definitions/SetPropertyOptions",
    },
    "type": {
      "const": "SET_PROPERTY",
      "examples": [
        "SET_PROPERTY",
      ],
      "type": "string",
    },
  },
  "required": [
    "options",
  ],
  "stability": "STABLE",
  "type": "object",
}
`;
